internal/xds/translator/httpfilters.go

```diff
@@ -140,16 +144,64 @@ func (o OrderedHTTPFilters) Swap(i, j int) {
 // For example, the cors filter should be put at the first to avoid unnecessary
 // processing of other filters for unauthorized cross-region access.
 // The router filter must be the last one since it's a terminal filter.
-func sortHTTPFilters(filters []*hcmv3.HttpFilter) []*hcmv3.HttpFilter {
+func sortHTTPFilters(filters []*hcmv3.HttpFilter, filterOrder []v1alpha1.FilterPosition) []*hcmv3.HttpFilter {
+	// Sort the filters in the default order.
 	orderedFilters := make(OrderedHTTPFilters, len(filters))
 	for i := 0; i < len(filters); i++ {
 		orderedFilters[i] = newOrderedHTTPFilter(filters[i])
 	}
 	sort.Sort(orderedFilters)
 
-	for i := 0; i < len(filters); i++ {
-		filters[i] = orderedFilters[i].filter
+	l := list.New()
+	for i := 0; i < len(orderedFilters); i++ {
+		l.PushBack(orderedFilters[i])
 	}
+
+	// Sort the filters in the custom order.
+	for i := 0; i < len(filterOrder); i++ {
```

Comment:
Is this a topological sorting problem? Can we use an established algorithm and library to create this sort? It would make it a bit easier to be convinced in the functional correctness.

Reply:
this logic could be simplified
1. init - ordered list
2. loop across filterOrder (overrideList)
3. find `name`, and delete it from list
4. find `before` or `after` and insert it before/after in list

Its still m (orig list)*n (override list) with both sizes being fairly small, but easier to comprehend

Reply:
> Is this a [topological sorting](https://en.wikipedia.org/wiki/Topological_sorting) problem? Can we use/reference an established algorithm (and possibly library) to create this sort? It would make it a bit easier to be convinced in the functional correctness.

There's no good way to combine two DAGs: the origin one and the custom one.  Would the current implementation be good enough since we have checked the cycle dependencies in Gateway API reconciler?